pkg/daemon/pinned_image_set_manager.go

```diff
@@ -0,0 +1,1096 @@
+func NewPinnedImageSetManager(
+	nodeName string,
+	criClient *cri.Client,
+	mcfgClient mcfgclientset.Interface,
+	imageSetInformer mcfginformersv1alpha1.PinnedImageSetInformer,
+	nodeInformer coreinformersv1.NodeInformer,
+	mcpInformer mcfginformersv1.MachineConfigPoolInformer,
+	minStorageAvailableBytes resource.Quantity,
+	runtimeEndpoint string,
+	authFilePath string,
+	prefetchTimeout time.Duration,
+	featureGatesAccessor featuregates.FeatureGateAccess,
+) *PinnedImageSetManager {
+	p := &PinnedImageSetManager{
+		nodeName:                 nodeName,
+		mcfgClient:               mcfgClient,
+		runtimeEndpoint:          runtimeEndpoint,
+		authFilePath:             authFilePath,
+		prefetchTimeout:          prefetchTimeout,
+		minStorageAvailableBytes: minStorageAvailableBytes,
+		featureGatesAccessor:     featureGatesAccessor,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pinned-image-set-manager"),
+		prefetchCh:               make(chan prefetch, defaultPrefetchWorkers*2),
+		criClient:                criClient,
+		backoff: wait.Backoff{
+			Steps:    maxRetries,
+			Duration: retryDuration,
+			Factor:   retryFactor,
+			Cap:      retryCap,
+		},
+	}
+
+	imageSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addPinnedImageSet,
+		UpdateFunc: p.updatePinnedImageSet,
+		DeleteFunc: p.deletePinnedImageSet,
+	})
+
+	mcpInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addMachineConfigPool,
+		UpdateFunc: p.updateMachineConfigPool,
+		DeleteFunc: p.deleteMachineConfigPool,
+	})
+
+	p.syncHandler = p.sync
+	p.enqueueMachineConfigPool = p.enqueue
+
+	p.imageSetLister = imageSetInformer.Lister()
+	p.imageSetSynced = imageSetInformer.Informer().HasSynced
+
+	p.nodeLister = nodeInformer.Lister()
+	p.nodeListerSynced = nodeInformer.Informer().HasSynced
+
+	p.mcpLister = mcpInformer.Lister()
+	p.mcpSynced = mcpInformer.Informer().HasSynced
+
+	return p
+}
+
+func (p *PinnedImageSetManager) sync(key string) error {
+	node, err := p.nodeLister.Get(p.nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get node %q: %w", p.nodeName, err)
+	}
+
+	pools, _, err := helpers.GetPoolsForNode(p.mcpLister, node)
+	if err != nil {
+		return err
+	}
+
+	if len(pools) == 0 {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), p.prefetchTimeout)
+	// cancel any currently running tasks in the worker pool
+	p.resetWorkload(cancel)
+
+	if err := p.syncMachineConfigPools(ctx, pools); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			ctxErr := fmt.Errorf("requeue: prefetching images incomplete after: %v", p.prefetchTimeout)
+			p.updateStatusError(ctxErr.Error())
+			klog.Error(ctxErr)
+			return ctxErr
+		}
+		p.updateStatusError(err.Error())
+		return err
+	}
+
+	p.updateStatusProgressingComplete("All pinned image sets complete")
```

Comment:
How would the controller (or whoever is responding to this) know which pinned image set is completed here?

e.g. if I set a pinned image set, and the daemons start pulling, and I modify that imageset to have more images, and a sub-section of nodes have completed the first pull, what would the state look like?

Reply:
Yep good question this is not completed yet but MCN status will report it accurately. the API has a current generation and a desired generation.

https://github.com/openshift/api/blob/abd990ce290b7e9481e33a1a71e045b39cc77ca1/machineconfiguration/v1alpha1/types_machineconfignode.go#L130-L154

Reply:
Ah good call, so at the highest level, the controller will reflect back onto the pinnedimageset object whether the latest generation has been completed by all the nodes?

Reply:
so MCN will provide detailed information on a node level and then the higher level status will be reported by MCP. 

our status will get plummed into the pools status reporting.
https://github.com/openshift/machine-config-operator/blob/d9ffefd68d2930918b27c4d030bf459943665023/pkg/controller/node/status.go#L92-L99

My hope was this would be enough but since the spec is on the pool also feels reasonable the status is on the pool?

Reply:
to follow up on this we have come to agreement on adding  https://github.com/openshift/api/pull/1852